Dataexample update

hrobarts opened the pull request:
## Changes
Use `zenodo_get` for the data download and validation

## Testing you performed
Updated download_data tests
- Mock zip file when `zenodo_get` is called
- Create the mock zip file in the CIL data folder rather than tmp folder because we don't always have access to unzip files in the tmp folder

## Related issues/links
depends on https://github.com/dvolgyes/zenodo_get/pull/25
## Checklist
- [x] I have performed a self-review of my code
- [x] I have added docstrings in line with the guidance in the developer guide
- [x] I have updated the relevant documentation
- [x] I have implemented unit tests that cover any new or modified functionality
- [x] CHANGELOG.md has been updated with any functionality change
- [x] Request review from all relevant developers
- [x] Change pull request label to 'Waiting for review'

## Contribution Notes
Please read and adhere to the [developer guide](https://tomographicimaging.github.io/CIL/nightly/developer_guide/) and local patterns and conventions.
- [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in CIL (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License
- [x] I confirm that the contribution does not violate any intellectual property rights of third parties

lauramurgatroyd commented on Wrappers/Python/test/test_dataexample.py:
```diff
@@ -157,116 +156,74 @@ def test_load_SIMULATED_CONE_BEAM_DATA(self):
-        '''
-        Test the _download_and_extract_data_from_url function correctly extracts files from a byte string
-        The zipped byte string is mocked using a temporary local zip file
+    def mock_zenodo_get(*args):
```
Possibly for each data we could also test that zenodo_get is being called with the correct zenodo record / zip file combo, but also that might be overkill 😄

hrobarts replied:
Good points, I have added test_zenodo_record_exists which tries to access the record and prints the available zip files, it will fail if the record can't be accessed or the zip file we want is not in the list. I think this addresses both points because it will fail if the record / zip file combo is wrong or not available

lauramurgatroyd replied:
Excellent 😄

lauramurgatroyd reviewed (changes requested):
It looks good, I have just suggested some changes to docs and I think the unit tests could be expanded a bit.

I think it would be nice to add further documentation with examples of how to use it. I just tested trying to get the walnut data locally.

I was expecting to be able to do dataexample.WALNUT.get but I think that was wrong - I had to use dataexample.WALNUT.download_data

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -295,8 +294,31 @@ class SANDSTONE(REMOTEDATA):
+    ZENODO_RECORD = '4912435'
+    ZIP_FILE = 'small.zip'
+
+    def get(cls, data_dir, filename):
```
```suggestion
    @classmethod
    def get(cls, data_dir, filename):
```

hrobarts replied:
Thanks! Updated now

hrobarts commented on scripts/requirements-test.yml:
```diff
@@ -1,16 +1,3 @@
-# Copyright 2023 United Kingdom Research and Innovation
-# Copyright 2023 The University of Manchester
-# Licensed under the Apache License, Version 2.0 (the "License");
```
Not sure how that happened, thank you for catching it!

hrobarts commented on CHANGELOG.md:
```diff
@@ -49,6 +49,7 @@
     - BlockOperator that would return a BlockDataContainer of shape (1,1) now returns the appropriate DataContainer. BlockDataContainer direct and adjoint methods accept DataContainer as parameter (#1802).
     - BlurringOperator: remove check for geometry class (old SIRF integration bug) (#1807)
     - The `ZeroFunction` and `ConstantFunction` now have a Lipschitz constant of 1. (#1768)
+    - Update dataexample remote data download to work with windows (#1774)
```
Good idea, updated now :)

hrobarts commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -295,8 +294,31 @@ class SANDSTONE(REMOTEDATA):
+    def get(cls, data_dir, filename):
+        '''
+        A microcomputed tomography dataset of a sunflower seeds in a box from https://zenodo.org/records/6874123
+        This function can be used to return selected projections or slices of the reconstruction
```
Thanks, updated now!

hrobarts commented on Wrappers/Python/test/test_dataexample.py:
```diff
@@ -157,116 +156,74 @@ def test_load_SIMULATED_CONE_BEAM_DATA(self):
+        Test the download_data function, when the user input is 'y' to 'are you sure you want to download data'
+        The user input to confirm the download is mocked as 'y'
+        The zip file download is mocked by creating a zip file locally
+        Test the download_data function correctly extracts files from the zip file
```
I've added a check that the zip file is deleted

hrobarts commented on Wrappers/Python/test/test_dataexample.py:
```diff
@@ -157,116 +156,74 @@ def test_load_SIMULATED_CONE_BEAM_DATA(self):
+        # redirect print output
+        capturedOutput = StringIO()                
+        sys.stdout = capturedOutput
+        for data in self.data_list:
```
Yes to check that download_data is inherited and also to check the folder can be set up correctly for each record - maybe it's overkill to do it for every dataset, especially if we add more datasets in future

hrobarts commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -295,8 +294,31 @@ class SANDSTONE(REMOTEDATA):
+        ----------
+        data_dir: str
+           The path to the directory where the dataset is stored. Data can be downloaded with dataexample.KORN.download_data(data_dir)
+
```
Thanks, updated now!

hrobarts commented on scripts/create_local_env_for_cil_development.sh:
```diff
@@ -64,6 +64,7 @@ conda_args=(create --name="$name"
   scipy
   tqdm
+  zenodo_get'>1.5.1'
 )
```
Yes maybe when we update the demos!

hrobarts commented:
> It looks good, I have just suggested some changes to docs and I think the unit tests could be expanded a bit.
> 
> I think it would be nice to add further documentation with examples of how to use it. I just tested trying to get the walnut data locally.
> 
> I was expecting to be able to do dataexample.WALNUT.get but I think that was wrong - I had to use dataexample.WALNUT.download_data

Thank you for the review Laura! You can just use dataexample.WALNUT.get(data_dir) if the data is already in data_dir but otherwise you have to use dataexample.WALNUT.download_data we decided it's best to be really explicit if we're downloading large datasets to the users computer. I agree this is not super clear so I will add some examples.

lauramurgatroyd commented:
> > It looks good, I have just suggested some changes to docs and I think the unit tests could be expanded a bit.
> > I think it would be nice to add further documentation with examples of how to use it. I just tested trying to get the walnut data locally.
> > I was expecting to be able to do dataexample.WALNUT.get but I think that was wrong - I had to use dataexample.WALNUT.download_data
> 
> Thank you for the review Laura! You can just use dataexample.WALNUT.get(data_dir) if the data is already in data_dir but otherwise you have to use dataexample.WALNUT.download_data we decided it's best to be really explicit if we're downloading large datasets to the users computer. I agree this is not super clear so I will add some examples.

Ok, that makes sense, but yes examples would be great!

hrobarts commented:
I've updated the docs to describe what the remote datasets are and how to use them:

![image](https://github.com/user-attachments/assets/65f3862e-8397-44a9-931c-e4fcc37116e2)

Also removed the test checking if the zenodo record exists.

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -46,19 +45,13 @@ def get(cls, size=None, scale=(0,1), **kwargs):
-                zipfile.extractall(path = data_dir)
-
     @classmethod
     def download_data(cls, data_dir):
```
maybe only ask for `input` if `prompt`?

```suggestion
    def download_data(cls, data_dir, prompt=True):
```

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -227,15 +232,21 @@ def get(cls, data_dir):
-    FOLDER = 'USB'
-    URL = 'https://zenodo.org/record/4822516/files/usb.zip'
-    FILE_SIZE = '3.2 GB'
+    FOLDER = 'USB' 
```
```suggestion
    FOLDER = 'USB'
```

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -293,10 +311,39 @@ class SANDSTONE(REMOTEDATA):
+        '''
+        extension = os.path.splitext(filename)[1]
+        if extension == '.mat':
+            return loadmat(os.path.join(data_dir,filename))
```
```suggestion
            return loadmat(os.path.join(data_dir,filename))
        raise KeyError(f"Unknown extension: {extension}")
```

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -71,12 +64,18 @@ def download_data(cls, data_dir):
-                print('Downloading dataset from ' + cls.URL)
-                cls._download_and_extract_from_url(os.path.join(data_dir,cls.FOLDER))
-                print('Download complete')
+            # get user confirmation for download
+            user_input = 'y' if not prompt else input("Are you sure you want to download {} dataset from Zenodo record {} ? (y/n)").format(cls.ZIP_FILE, cls.ZENODO_RECORD)
+            if  user_input == "y": 
```
```suggestion
            user_input = input("Are you sure you want to download {cls.ZIP_FILE} dataset from Zenodo record {cls.ZENODO_RECORD}? [Y/n]: ") if prompt else 'y'
            if  user_input.lower() == 'y':
```

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -71,12 +64,18 @@ def download_data(cls, data_dir):
+                with ZipFile(os.path.join(data_dir, cls.ZIP_FILE), 'r') as zip_ref:
+                    zip_ref.extractall(os.path.join(data_dir, cls.FOLDER))
+                os.remove(os.path.join(data_dir, cls.ZIP_FILE))
+
             else:
                 print('Download cancelled')
```
```suggestion
            return True
            else:
                print('Download cancelled')
                return False
```

might be useful to return success as a `bool` for end-user scripts...

casperdcl commented on Wrappers/Python/cil/utilities/dataexample.py:
```diff
@@ -71,14 +64,20 @@ def download_data(cls, data_dir):
-                cls._download_and_extract_from_url(os.path.join(data_dir,cls.FOLDER))
-                print('Download complete')
+            user_input = input("Are you sure you want to download {cls.ZIP_FILE} dataset from Zenodo record {cls.ZENODO_RECORD}? [Y/n]: ") if prompt else 'y'
+            if  user_input.lower() == 'y':
+                zenodo_get([cls.ZENODO_RECORD, '-g', cls.ZIP_FILE, '-o', data_dir])
+
+                # unzip file
+                with ZipFile(os.path.join(data_dir, cls.ZIP_FILE), 'r') as zip_ref:
+                    zip_ref.extractall(os.path.join(data_dir, cls.FOLDER))
+                os.remove(os.path.join(data_dir, cls.ZIP_FILE))
             else:
                 print('Download cancelled')
+                return False
+            return True
```
```suggestion
            if user_input.lower() not in ('y', 'yes'):
                print('Download cancelled')
                return False

            zenodo_get([cls.ZENODO_RECORD, '-g', cls.ZIP_FILE, '-o', data_dir])
            with ZipFile(os.path.join(data_dir, cls.ZIP_FILE), 'r') as zip_ref:
                zip_ref.extractall(os.path.join(data_dir, cls.FOLDER))
            os.remove(os.path.join(data_dir, cls.ZIP_FILE))
            return True
```

slight refactor to remove most indents...
